drivers/gpio/gpio_silabs_siwx91x.c

```diff
@@ -158,6 +136,67 @@ static int gpio_siwx91x_pin_configure(const struct device *dev, gpio_pin_t pin,
 
 	sl_si91x_gpio_set_pin_direction(cfg->hal_port, pin, (flags & GPIO_OUTPUT) ? 0 : 1);
 
+	for (i = 0; i < data->next_pin; i++) {
+		if (data->pin_config_info[i].pin == pin) {
+			cur_cfg_pin = i;
+			break;
+		}
+	}
+
+	if (i == data->next_pin) {
+		cur_cfg_pin = data->next_pin;
+		data->next_pin++;
+	}
+
+	if (cur_cfg_pin < __builtin_popcount(cfg->common.port_pin_mask)) {
+		data->pin_config_info[cur_cfg_pin].port_dev = dev;
+		data->pin_config_info[cur_cfg_pin].pin = pin;
+		data->pin_config_info[cur_cfg_pin].flags = flags;
+	} else {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int gpio_siwx91x_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct gpio_siwx91x_common_config *pcfg = dev->config;
+	struct gpio_siwx91x_common_data *pdata = dev->data;
+	const struct device **port_dev = pdata->ports;
+	const struct gpio_siwx91x_port_config *port_cfg = NULL;
+	struct gpio_siwx91x_port_data *port_data = NULL;
+	int pin_cnt;
+	int ret;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
+		for (int i = 0; i < pcfg->port_count; i++) {
+			port_cfg = port_dev[i]->config;
+			port_data = port_dev[i]->data;
+			pin_cnt = 0;
+			while (port_data->pin_config_info[i].port_dev != NULL &&
+			       pin_cnt < __builtin_popcount(port_cfg->common.port_pin_mask)) {
+				ret = gpio_siwx91x_pin_configure(
+					port_data->pin_config_info[pin_cnt].port_dev,
+					port_data->pin_config_info[pin_cnt].pin,
+					port_data->pin_config_info[pin_cnt].flags);
```

Comment:
Is it really correct that you're using a different index for `port_data->pin_config_info` in the while loop condition and inside the loop? Looks a bit suspicious.